cudax/include/cuda/experimental/__utility/basic_any/semiregular.cuh

```diff
@@ -145,37 +147,106 @@ struct icopyable : interface<icopyable, extends<imovable<>>>
   }
 };
 
-template <class...>
-struct iequality_comparable : interface<iequality_comparable>
+template <class... _Super>
+struct iequality_comparable;
+
+struct iequality_comparable_base : interface<iequality_comparable>
 {
-  _CCCL_TEMPLATE(class _Tp)
-  _CCCL_REQUIRES(_CUDA_VSTD::equality_comparable<_Tp>)
-  using overrides _CCCL_NODEBUG_ALIAS = overrides_for<_Tp, _CUDAX_FNPTR_CONSTANT_WAR(&__equal_fn<_Tp>)>;
+  // These overloads are only necessary so that iequality_comparable<> itself
+  // satisfies the std::equality_comparable constraint that is used by the
+  // `iequality_comparable<>::overloads` alias template below.
+  _CCCL_NODISCARD_FRIEND _CUDAX_TRIVIAL_HOST_API auto
+  operator==(iequality_comparable<> const&, iequality_comparable<> const&) noexcept -> bool
+  {
+    _CCCL_UNREACHABLE();
+  }
 
-#if !defined(_CCCL_NO_THREE_WAY_COMPARISON)
-  _CCCL_NODISCARD _CUDAX_HOST_API auto operator==(iequality_comparable const& __other) const -> bool
+  _CCCL_NODISCARD_FRIEND _CUDAX_TRIVIAL_HOST_API auto
+  operator!=(iequality_comparable<> const&, iequality_comparable<> const&) noexcept -> bool
   {
-    auto const& __rhs = __cudax::basic_any_from(__other);
-    void const* __obj = __basic_any_access::__get_optr(__rhs);
-    return __cudax::virtcall<&__equal_fn<iequality_comparable>>(this, __rhs.type(), __obj);
+    _CCCL_UNREACHABLE();
   }
-#else // ^^^ !_CCCL_NO_THREE_WAY_COMPARISON ^^^ / vvv _CCCL_NO_THREE_WAY_COMPARISON vvv
+
+  template <class _Interface>
+  struct __object
+  {
+    _CCCL_TEMPLATE(class _Object)
+    _CCCL_REQUIRES(__satisfies<_Object, _Interface>)
+    __object(_Object const& __obj) noexcept
+        : __obj_(&__obj)
+        , __type_(_CCCL_TYPEID(_Object))
+    {}
+
+    const void* __obj_;
+    _CUDA_VSTD::__type_info_ref __type_;
+  };
+
+  // These are the overloads that actually get used when testing `basic_any`
+  // objects for equality.
+  _CCCL_TEMPLATE(class _ILeft, class _IRight)
+  _CCCL_REQUIRES(_CUDA_VSTD::convertible_to<basic_any<_ILeft> const&, basic_any<_IRight> const&>
+                 || _CUDA_VSTD::convertible_to<basic_any<_IRight> const&, basic_any<_ILeft> const&>)
```

Comment:
In the implementation this is only converting __rhs.

On the other hand it passes a pointer to the virtcall so that is the shortcut you meant?.

The convertible_to only checks whether we could construct the virtcall and then pretents that we did?